server/datastore/mysql/migrations/tables/20231212094238_AddUniqueHashToSoftware.go

```diff
@@ -0,0 +1,72 @@
+package tables
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20231212094238, Down_20231212094238)
+}
+
+func Up_20231212094238(tx *sql.Tx) error {
+	// binary(16) is the efficient way to store md5 hashes,
+	// see https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html
+	// We store it using UNHEX(MD5(<the string value to hash>)).
+	//
+	// We use md5 even though it is insecure because it is used only to
+	// deduplicate software entries, not for any security purpose (i.e. a
+	// maliciously-crafted collision would not cause serions issues, and the
+	// source of input is osquery-extracted software data), and we already use it
+	// for configuration profiles so instead of using different hashing
+	// algorithms, we'll stick to md5.
+	_, err := tx.Exec(`ALTER TABLE software ADD COLUMN checksum BINARY(16) DEFAULT NULL`)
+	if err != nil {
+		return fmt.Errorf("failed to add checksum column to software table: %w", err)
+	}
+
+	// fill the checksum for existing rows - order of column used to generate the
+	// checksum is important, we will need to use the same everywhere. The logic
+	// of that computed checksum is captured in
+	// mysql.softwareChecksumComputedColumn, but we don't use it here because if
+	// the function's implementation changes in the future, it should not affect
+	// this DB migration (e.g. the function might use columns that don't exist at
+	// the point in time when this migration is run).
+	_, err = tx.Exec(`
+UPDATE
+	software
+SET
+	checksum = UNHEX(
+		MD5(
+			-- concatenate with separator \x00
+			CONCAT_WS(CHAR(0),
+				name,
+				version,
+				source,
+				COALESCE(bundle_identifier, ''),
```

Comment:
Why COALESCE used just here and not used other places?

Reply:
This is the only NULL-able field in the table.